[MRG] Vectorizer class to help chain MNE transformers by converting o/p into 2D

kaichogami opened the pull request:
This was previously on #3276. The idea is to support MNE transformers with scikit-learn pipeline and be compatible with `cross_val_score` and `grid_search` methods. Ideally this should be in the last step in a pipeline to convert higher dimension output into two dimension.
Example
```
clf = make_pipeline(SptialFilter(), XdawnTransformer(), Vectorizer(), LogisticRegression())
cross_val_score(clf)
```

kaichogami commented:
@Eric89GXL Why is this failing here? I created this branch after updating with master.

mmagnuski commented:
The branch does not have conflicts with master anymore so it is ok.
Travis is complaining because of this error:
```
======================================================================
ERROR: Test module docsting formatting
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/miniconda/envs/testenv/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/travis/build/mne-tools/mne-python/mne/utils.py", line 733, in dec
    return function(*args, **kwargs)
  File "/home/travis/build/mne-tools/mne-python/mne/tests/test_docstring_parameters.py", line 121, in test_docstring_parameters
    % (cls, name, w[0]))
RuntimeError: Error for __init__ of <class 'mne.decoding.transformer.Vectorizer'> in mne.decoding:
{message : UserWarning('Unknown section Example',), category : 'UserWarning', filename : '/home/travis/miniconda/envs/testenv/lib/python2.7/site-packages/numpydoc/docscrape.py', lineno : 136, line : None}
```

The Example section is not recognized, it should be `Examples`.

kaichogami commented:
@mmagnuski Thank you for pointing that out!

agramfort commented on mne/decoding/tests/test_transformer.py:
```diff
@@ -160,3 +160,17 @@ def test_epochs_vectorizer():
+                       eog=False, exclude='bads')
+    picks = picks[1:13:3]
+    epochs = Epochs(raw, events, event_id, tmin, tmax, picks=picks,
+                    baseline=(None, 0), preload=True)
```
you don't need to go via all this just to test the vectorizer on a 3d array. Just use a random 3d array.

agramfort commented on mne/decoding/transformer.py:
```diff
@@ -245,6 +245,75 @@ def inverse_transform(self, X, y=None):
+        """
+        return self
+
+    def transform(self, X):
```
an inverse_transform is also needed

agramfort commented:
this is a small PR which is good but I have hard time getting the big picture as it's just new code and appears in no example or in the narrative doc.

kaichogami commented:
@agramfort I added an introductory doc. Does it look good enough?

kingjr commented on doc/manual/decoding.rst:
```diff
@@ -27,6 +27,10 @@ Scikit-learn API enforces the requirement that data arrays must be 2D. A common
 
+Vectorizer
+^^^^^^^^^^
+Scikit-learn API provides functionality to chain transformers and estimators by using :class:`sklearn.Pipeline`. We can apply MNE transformers in scikit-learn pipeline to chain them. Further it becomes useful for cross validation and grid search for parameters. Vectorizer is typically the last step(before the estimator in the pipeline) to convert higher dimensional data into two dimensional, comp ...
```
'sklearn.Pipeline' => 'sklearn.pipeline.Pipeline'

'Further' => 'Furthermore'

'step(' => 'step ('

'cross validation' => 'cross-validation'
'grid search'  => 'grid-search'

Overall, check your English, the current sentences are not really clear/specific. I would write something like:

"Decoding pipelines may be constructed following Scikit-Learn API. However, whereas Scikit-Learn transformers and estimators generally expect 2D data (`n_samples` * `n_features`), MNE transformers typically output data with higher dimensionality (e.g. `n_samples` * `n_channels` * `n_frequencies` * `n_times`). A `Vectorizer` therefore needs to be applied between the MNE and the Scikit-Learn steps: e.g: `make_pipeline(Xdawn(), Vectorizer(), LogisticRegression())`

kingjr commented:
@kaichogami How come there isn't any red lines? Doesn't `Vectorizer` already exist?

@agramfort sklearn estimators and transformers expect 2D arrays, whereas most of MNE transformer will output 3 or 4 dimensional `X`: e.g. `Xdawn().fit_transform(X, y)` generates a `n_trials * n_components * n_times` `Xt`. If we want to apply a logistic, we first need to vectorize the output.

I believe that the `Xdawn` example which uses Vectorizer will be updated once the corresponding PR will be merged; @kaichogami is that right?

agramfort commented:
then let's merge this after xdawn to fully understand the use case.

kaichogami commented:
@kingjr, what do you mean by red lines? We didn't create `Vectorizer`.
@agramfor @kingjr yes, example would only make sense after `Vectorizer` is merged.

kingjr commented:
@kaichogami can you rebase this branch? Also, if it applies, can you update the xdawn example to use this vectorizer, (or else explain why it is not necessary)

kaichogami commented on examples/decoding/plot_decoding_xdawn_transformer.py:
```diff
@@ -17,7 +17,7 @@
 from mne import io, pick_types, read_events, Epochs
 from mne.datasets import sample
-from mne.decoding.xdawn import XdawnTransformer
+from mne.decoding import Vectorizer, XdawnTransformer
```
@agramfort I corrected the import statement from #3245 here.

kingjr commented on mne/decoding/transformer.py:
```diff
@@ -245,6 +245,92 @@ def inverse_transform(self, X, y=None):
+        ----------
+        X : numpy array, shape(n_epochs, n_chans, n_times) or
+                              (n_epochs, n_chans * n_times) or
+                              (n_epochs, n_chans, n_times, n_freqs)
```
We need need to specify all cases, X just need to have `n_epochs` in its first dimension.

You can specify: 
`
X : array-like
    The data to fit. Can be, for example a list, or an array at least 2d. The first dimension must be of shape `n_sample`.
`

mmagnuski replied:
That would be better, especially as time-frequency is in `(n_epochs, n_chans, n_freqs, n_times)` in mne, not the way it is described above.

kingjr commented on mne/decoding/transformer.py:
```diff
@@ -245,6 +245,92 @@ def inverse_transform(self, X, y=None):
+        self : Instance of Vectorizer
+            Return the modified instance.
+        """
+        self.shape = X.shape
```
self.shape_

kingjr commented on mne/decoding/transformer.py:
```diff
@@ -245,6 +245,92 @@ def inverse_transform(self, X, y=None):
 
 
+class Vectorizer(TransformerMixin):
+    """Class to chain MNE transformer output to scikit-learn estimators.
```
Change it to 'Transforms n-dimensional array into 2D array of n_sample by n_features.'

kingjr commented on mne/decoding/transformer.py:
```diff
@@ -245,6 +245,92 @@ def inverse_transform(self, X, y=None):
+
+    Examples
+    -------
+    clf = make_pipeline(SpatialFilter(), XdawnTransforme(), Vectorizer(),
```
XdawnTransformer

kingjr commented on mne/decoding/transformer.py:
```diff
@@ -245,6 +245,92 @@ def inverse_transform(self, X, y=None):
+    -------
+    clf = make_pipeline(SpatialFilter(), XdawnTransforme(), Vectorizer(),
+                        LogisticRegression())
+    cross_val_score(clf)
```
remove cross_val_score

kingjr commented:
@Eric89GXL The travis error seem unrelated to this PR. Do you have an idea of how we can debug this?

Eric89GXL commented:
I should be able to fix it in a couple of hours

kingjr commented:
Thanks!

mmagnuski commented on mne/decoding/transformer.py:
```diff
@@ -245,6 +245,92 @@ def inverse_transform(self, X, y=None):
+        ----------
+        X : numpy array, shape(n_epochs, n_chans, n_times) or
+                              (n_epochs, n_chans * n_times)
+                              (n_epochs, n_chans, n_times, n_freqs)
```
same comment here - time-freq is in `(n_epochs, n_chans, n_times, n_freqs)` or `(n_chans, n_times, n_freqs)`

Eric89GXL commented:
Can you rebase to `master`? It should fix Travis. Then is this good to go @kingjr? We should get rid of the CircleCI `master` failure ASAP, so either this should go in or we should do it in a different PR.

kingjr commented on doc/manual/decoding.rst:
```diff
@@ -27,6 +27,10 @@ Scikit-learn API enforces the requirement that data arrays must be 2D. A common
 
+Vectorizer
+^^^^^^^^^^
+Scikit-learn API provides functionality to chain transformers and estimators by using :class:`sklearn.pipeline.Pipeline`. We can construct decoding pipeline and perform cross-validation and grid-search. However scikit-Learn transformers and estimators generally expect 2D data (n_samples * n_features), MNE transformers typically output data with higher dimensionality (e.g. n_samples * n_channels * ...
```
', MNE transformers' => ', whereas MNE transformers'

kingjr commented on doc/whats_new.rst:
```diff
@@ -101,6 +101,8 @@ API
 
     - Added :class:`mne.decoding.XdawnTransformer` implementing xdawn filtering algorithm working with numpy arrays of epochs data and compatible with scikit-learn pipeline. Does not support overlap case. By `Asish Panda`_
 
+    - Added :class:`mne.decoding.Vectorizer` to convert MNE transformer output into two dimension as required in scikit-learn pipeline. By `Asish Panda`_
```
"two dimensions"

kingjr commented on examples/decoding/plot_decoding_xdawn_transformer.py:
```diff
@@ -40,9 +40,10 @@
 
 X = epochs.get_data()
-y = label_binarize(epochs.events[:, 2], classes=[1, 3])
+y = label_binarize(epochs.events[:, 2], classes=[1, 3]).reshape(len(X))
```
why reshape?

kingjr replied:
it is weirdly hacky, what is the shape of y after label_binarize?

kaichogami replied:
`y` becomes a single column vector which resulted in the error @Eric89GXL mentioned in #3433

kingjr commented on mne/decoding/tests/test_transformer.py:
```diff
@@ -160,3 +160,13 @@ def test_epochs_vectorizer():
+    result = vect.fit_transform(data)
+    assert_equal(result.ndim, 2)
+    # check inverse_trasnform
+    assert_equal(vect.inverse_transform(result).ndim, 3)
```
add 
`assert_equal_array(vect.inverse_transform(result), data)`

kingjr replied:
add a test to check that you can pass different number of samples:
```
vect.fit(data)
vect.transform(data[1:])
```

agramfort replied:
still to be done

kingjr replied:
Additionally, check that you can't pass an array with a different number of features with an `assert_raise`.

kingjr commented on mne/decoding/transformer.py:
```diff
@@ -245,6 +245,92 @@ def inverse_transform(self, X, y=None):
+class Vectorizer(TransformerMixin):
+    """Transforms n-dimensional array into 2D array of n_sample by n_features.
+
+    MNE transformer have three, four dimensional output. This class converts
```
Remove 'MNE transformer have ... output.", and rephrase the second second sentence into:

This class reshapes an n-dimensional array into an n_sample * n_features array, usable by the estimators and transformers of scikit-learn.

kingjr commented on mne/decoding/transformer.py:
```diff
@@ -245,6 +245,92 @@ def inverse_transform(self, X, y=None):
+    MNE transformer have three, four dimensional output. This class converts
+    them into two dimension so as to comply with scikit-learn API.
+
+    Examples
```
Example

kaichogami replied:
Writing `Example` results in error while building the doc.

kingjr commented on mne/decoding/transformer.py:
```diff
@@ -245,6 +245,92 @@ def inverse_transform(self, X, y=None):
+    """
+
+    def fit(self, X, y=None):
+        """Does nothing. Added for scikit-learn compatibility.
```
Store the shape of X.

kingjr commented on mne/decoding/transformer.py:
```diff
@@ -245,6 +245,92 @@ def inverse_transform(self, X, y=None):
+
+    MNE transformer have three, four dimensional output. This class converts
+    them into two dimension so as to comply with scikit-learn API.
+
```
You need to add something like: 

```
Attribute:
------------
shape : array
  The original dimension of the data.
```

kingjr commented on mne/decoding/transformer.py:
```diff
@@ -245,6 +245,92 @@ def inverse_transform(self, X, y=None):
+
+        Returns
+        -------
+        X : numpy ndarray of shape(n_trials, n_chan * n_times * n_freqs)
```
correct this shape

kingjr commented on mne/decoding/transformer.py:
```diff
@@ -245,6 +245,92 @@ def inverse_transform(self, X, y=None):
+        X : array-like
+            The data to be transformed. Can be, for example a list, or an 
+            array of at least 2d. The first dimension must be of
+            shape (n_sample).
```
n_samples, same everywhere

kingjr commented on mne/decoding/transformer.py:
```diff
@@ -245,6 +245,92 @@ def inverse_transform(self, X, y=None):
+
+        Returns
+        -------
+        X : numpy ndarray of shape(n_trials, n_chan * n_times * n_freqs)
```
remove numpy in dodcs, just indicate `array, shape(...`

kingjr commented on mne/decoding/transformer.py:
```diff
@@ -245,6 +245,92 @@ def inverse_transform(self, X, y=None):
+        return self.fit(X).transform(X)
+
+    def inverse_transform(self, X):
+        """Transform 2D data back to shape used in fit.
```
back to its original feature shape.

kingjr commented on mne/decoding/transformer.py:
```diff
@@ -245,6 +245,92 @@ def inverse_transform(self, X, y=None):
+            The data transformed into shape as used in fit. The first
+            dimension is of shape (n_sample).
+        """
+        return X.reshape(self.shape_)
```
This is incorrect, should be `X.reshape(np.r_[len(X), self.shape_])`

kingjr commented on mne/decoding/transformer.py:
```diff
@@ -245,6 +245,92 @@ def inverse_transform(self, X, y=None):
+        self : Instance of Vectorizer
+            Return the modified instance.
+        """
+        self.shape_ = X.shape
```
don't store the first dimension, it can very across fit/transform.

`self.shape_ = X.shape[1:]`

kaichogami commented:
@kingjr @dengemann @mmagnuski @agramfort comments?

agramfort commented on mne/decoding/transformer.py:
```diff
@@ -245,6 +245,97 @@ def inverse_transform(self, X, y=None):
+        X : array-like
+            The data to be transformed. Can be, for example a list, or an
+            array of at least 2d. The first dimension must be of
+            shape (n_samples).
```
The first dimension must be of shape (n_samples).
->
The first dimension must be of length n_samples.

fix below too

agramfort commented on mne/decoding/transformer.py:
```diff
@@ -245,6 +245,97 @@ def inverse_transform(self, X, y=None):
+            The data transformed into shape as used in fit. The first
+            dimension is of shape (n_samples).
+        """
+        return X.reshape(np.r_[len(X), self.shape_])
```
X.reshape((len(X),) + self.shape_)

you can use + for tuples

agramfort commented:
@kaichogami set PR to MRG when all comments are addressed.

mmagnuski commented on mne/decoding/transformer.py:
```diff
@@ -245,6 +245,97 @@ def inverse_transform(self, X, y=None):
 
 
+class Vectorizer(TransformerMixin):
+    """Transforms n-dimensional array into 2D array of n_sample by n_features.
```
n_sample**s**, the same below

mmagnuski commented on mne/decoding/transformer.py:
```diff
@@ -245,6 +245,97 @@ def inverse_transform(self, X, y=None):
+    Attributes
+    ----------
+    ``shape_`` : array
+         Stores the original dimension of data.
```
I think it would be better either dimension**s** or just shape

mmagnuski replied:
BTW - I think shape would be a tuple, not array

kingjr replied:
I agree with @mmagnuski. Also change `shape_` to `features_shape_` to be more explicit.

kingjr commented on doc/manual/decoding.rst:
```diff
@@ -27,6 +27,10 @@ Scikit-learn API enforces the requirement that data arrays must be 2D. A common
 
+Vectorizer
+^^^^^^^^^^
+Scikit-learn API provides functionality to chain transformers and estimators by using :class:`sklearn.pipeline.Pipeline`. We can construct decoding pipeline and perform cross-validation and grid-search. However scikit-Learn transformers and estimators generally expect 2D data (n_samples * n_features), whereas MNE transformers typically output data with higher dimensionality (e.g. n_samples * n_ch ...
```
'We can construct a decoding pipeline' or 'We can construct decoding pipelines'

'with a higher dimensionality'. 

'scikit-learn' (not upper case)

kingjr commented on doc/whats_new.rst:
```diff
@@ -107,6 +107,8 @@ API
 
     - A new option ``set_env`` has been added to :func:`mne.set_config` that defaults to ``False`` in 0.13 but will change to ``True`` in 0.14, by `Eric Larson`_
 
+    -  Added :class:`mne.decoding.Vectorizer` to convert MNE transformer output into two dimensions as required in scikit-learn pipeline. By `Asish Panda`_
```
to convert the output of an MNE transformer into two dimensions so as to pipeline it to scikit-learn estimators.

mmagnuski commented on mne/decoding/transformer.py:
```diff
@@ -245,6 +245,97 @@ def inverse_transform(self, X, y=None):
+        return self
+
+    def transform(self, X):
+        """Convert matrix data into two dimensions.
```
Better not to use matrix here just in case someone confuses it with `np.matrix` which is always 2D.

kingjr commented on mne/decoding/transformer.py:
```diff
@@ -245,6 +245,97 @@ def inverse_transform(self, X, y=None):
+    """
+
+    def fit(self, X, y=None):
+        """Stores the shape of X.
```
Stores the shape of the features of X.

kingjr commented on mne/decoding/transformer.py:
```diff
@@ -245,6 +245,97 @@ def inverse_transform(self, X, y=None):
+        self : Instance of Vectorizer
+            Return the modified instance.
+        """
+        self.shape_ = X.shape[1:]
```
add X = np.asarray(X) before, since your doc suggests it can be a list.

kingjr commented on mne/decoding/transformer.py:
```diff
@@ -245,6 +245,97 @@ def inverse_transform(self, X, y=None):
+        ----------
+        X : array-like
+            The data to fit. Can be, for example a list, or an array of at
+            least 2d. The first dimension must be of shape (n_samples).
```
must be of length n_samples.

mmagnuski replied:
Although `n_samples` is used everywhere in sklearn, here - in mne-python it might be confused with timesamples. I think it is worth some clarification just in case.

kingjr replied:
yes, althought vectorizer will be generic, and thus applicable to epoched, or raw data. @kaichogami  may be write 'of length n_samples, where samples are the independent samples used by the estimator (e.g. n_epochs for epoched data).

kingjr commented on mne/decoding/transformer.py:
```diff
@@ -245,6 +245,97 @@ def inverse_transform(self, X, y=None):
+
+        Returns
+        -------
+        X : array, shape (n_trials, n_chan * n_times)
```
shape (n_samples, -1)

kingjr replied:
same below

kingjr commented on doc/whats_new.rst:
```diff
@@ -113,6 +113,8 @@ API
 
     - The ``compensation`` parameter in :func:`mne.io.read_raw_fif` has been deprecated in favor of the method :meth:`mne.io.Raw.apply_gradient_compensation` by `Eric Larson`_
 
+    -   -  Added :class:`mne.decoding.Vectorizer` to convert the output of an MNE transformer into two dimensions so as to pipeline it to scikit-learn estimators. By `Asish Panda`_
```
no dot between estimators and by, see above

kingjr commented on mne/decoding/tests/test_transformer.py:
```diff
@@ -160,3 +160,17 @@ def test_epochs_vectorizer():
+    assert_array_equal(vect.inverse_transform(result), data)
+    # check with different shape
+    vect.fit_transform(np.random.rand(150, 18, 6, 3))
+    vect.fit_transform(data[1:])
```
still need to add a test to check that you can't pass an array with a different number of features with an assert_raise.

kingjr commented on mne/decoding/transformer.py:
```diff
@@ -245,6 +245,102 @@ def inverse_transform(self, X, y=None):
+        -------
+        self : Instance of Vectorizer
+            Return the modified instance.
+        """
```
add `X = np.asarray(X)`

kingjr commented on mne/decoding/transformer.py:
```diff
@@ -245,6 +245,102 @@ def inverse_transform(self, X, y=None):
+        -------
+        X : array, shape (n_trials, -1)
+            The transformed data.
+        """
```
add `X = np.asarray(X)`

kingjr commented on mne/decoding/transformer.py:
```diff
@@ -245,6 +245,102 @@ def inverse_transform(self, X, y=None):
+        -------
+        X : array, shape (n_trials, -1)
+            The transformed data.
+        """
```
add `X = np.asarray(X)`

kingjr commented on mne/decoding/transformer.py:
```diff
@@ -245,6 +245,102 @@ def inverse_transform(self, X, y=None):
+        X : array-like
+            The data transformed into shape as used in fit. The first
+            dimension is of length (n_samples).
+        """
```
add `X = np.asarray(X)`

kingjr commented:
Final small comments to be addressed, then ready to merge on my end.

mmagnuski commented on mne/decoding/transformer.py:
```diff
@@ -245,6 +245,102 @@ def inverse_transform(self, X, y=None):
+            The data to fit. Can be, for example a list, or an array of at
+            least 2d. The first dimension must be of length n_samples, where
+            samples are the independent samples used by the estimator
+            (e.g. n_epochs for epoched data).
```
I like it much better now, thanks! :)

mmagnuski commented on mne/decoding/transformer.py:
```diff
@@ -245,6 +245,102 @@ def inverse_transform(self, X, y=None):
+
+    Attributes
+    ----------
+    ``shape_`` : array
```
This is called `features_shape_` now, and you are only doing `X.shape[1:]` so it's a tuple, not array.

kaichogami commented:
@kingjr I raised errors in case of different shape of data used in fit and transform, with tests. Is this what you meant?

agramfort commented:
@kaichogami the number of samples between fit and transform can be different. So only X.shape[0] can vary but not the rest ok?

kaichogami commented:
@agramfort yes I have raised errors based on that condition.

kingjr commented:
@kaichogami yes. LGTM, thanks @kaichogami

agramfort commented on mne/decoding/tests/test_transformer.py:
```diff
@@ -160,3 +160,25 @@ def test_epochs_vectorizer():
+    # check inverse_trasnform
+
+    assert_equal(vect.inverse_transform(result).ndim, 3)
+    assert_array_equal(vect.inverse_transform(result), data)
```
don't run twice vect.inverse_transform(result)

agramfort commented on mne/decoding/tests/test_transformer.py:
```diff
@@ -160,3 +160,25 @@ def test_epochs_vectorizer():
+    assert_array_equal(vect.inverse_transform(result), data)
+
+    # check with different shape
+    vect.fit_transform(np.random.rand(150, 18, 6, 3))
```
check sizes of the output

agramfort commented on mne/decoding/tests/test_transformer.py:
```diff
@@ -160,3 +160,25 @@ def test_epochs_vectorizer():
+
+    # check with different shape
+    vect.fit_transform(np.random.rand(150, 18, 6, 3))
+    vect.fit_transform(data[1:])
```
here too

agramfort commented on mne/decoding/transformer.py:
```diff
@@ -245,6 +245,111 @@ def inverse_transform(self, X, y=None):
         return X.reshape(-1, self.n_channels, self.n_times)
 
 
+class Vectorizer(TransformerMixin):
```
did you deprecate EpochsVectorizer?

kaichogami replied:
@agramfort, I added a deprecate warning and replaced `EpochsVectorizer` with `Vectorizer` in the examples it was being used.

kingjr replied:
You have to add `@deprecated` (in mne.utils.deprecated IIRC), look across the package to see how to use it.

kaichogami replied:
Yep did that!

agramfort commented on mne/decoding/transformer.py:
```diff
@@ -147,6 +147,7 @@ def inverse_transform(self, epochs_data, y=None):
         return X
 
 
+@deprecated("EpochsVectorizer will be deprecated; use Vectorizer instead")
```
say in what version you'll deprecate. It will in 0.14

agramfort commented on mne/decoding/transformer.py:
```diff
@@ -245,6 +246,111 @@ def inverse_transform(self, X, y=None):
+
+        Returns
+        -------
+        X : array, shape (n_trials, -1)
```
n_trials -> n_samples

agramfort commented on mne/decoding/transformer.py:
```diff
@@ -245,6 +246,111 @@ def inverse_transform(self, X, y=None):
+
+        Parameters
+        ----------
+        X : array-like
```
shape?

agramfort replied:
I think this docstring is wrong. For inverse_transform X is 2D (n_samples, n_features)

please check your docstrings

agramfort commented on mne/decoding/transformer.py:
```diff
@@ -245,6 +246,111 @@ def inverse_transform(self, X, y=None):
+
+        Returns
+        -------
+        X : array-like
```
shape?

agramfort commented on mne/decoding/transformer.py:
```diff
@@ -245,6 +246,111 @@ def inverse_transform(self, X, y=None):
+
+        Returns
+        -------
+        X : array, shape (n_trials, -1)
```
n_samples

agramfort commented on mne/decoding/transformer.py:
```diff
@@ -245,6 +246,111 @@ def inverse_transform(self, X, y=None):
+
+        Parameters
+        ----------
+        X : array-like
```
shape?

agramfort commented on mne/decoding/transformer.py:
```diff
@@ -245,6 +246,111 @@ def inverse_transform(self, X, y=None):
+
+        Parameters
+        ----------
+        X : array-like
```
shape?

agramfort commented on mne/decoding/transformer.py:
```diff
@@ -245,6 +246,111 @@ def inverse_transform(self, X, y=None):
+
+        Parameters
+        ----------
+        X : array-like
```
shape?

kaichogami replied:
@agramfort Since the array could be of multiple shapes, how should it be written? This is similar to how scikit-learn [doc](https://github.com/scikit-learn/scikit-learn/blob/51a765a/sklearn/cross_validation.py#L1362) is written

agramfort replied:
ok ignore

agramfort commented:
did you "git grep EpochsVectorizer" to see if you removed all occurences except in whats_new?

kaichogami commented:
@agramfort yes did that!

agramfort commented on doc/whats_new.rst:
```diff
@@ -113,6 +113,8 @@ API
     - The ``compensation`` parameter in :func:`mne.io.read_raw_fif` has been deprecated in favor of the method :meth:`mne.io.Raw.apply_gradient_compensation` by `Eric Larson`_
 
+    - Added :class:`mne.decoding.Vectorizer` to convert the output of an MNE transformer into two dimensions so as to pipeline it to scikit-learn estimators by `Asish Panda`_
+
```
say that EpochsVectorizer is deprecated and replaced by Vectorizer. See how it was done below

agramfort replied:
@kaichogami please see here all unaddressed comments. Come on try to be reactive so we can iterate faster. These are 20 mins of your time.

kingjr commented on mne/decoding/transformer.py:
```diff
@@ -245,6 +246,111 @@ def inverse_transform(self, X, y=None):
+            least 2d. The first dimension must be of length n_samples, where
+            samples are the independent samples used by the estimator
+            (e.g. n_epochs for epoched data).
+        y : None
```
None | array, shape (n_samples,)

kingjr commented on mne/decoding/transformer.py:
```diff
@@ -245,6 +246,111 @@ def inverse_transform(self, X, y=None):
+        Returns
+        -------
+        X : array-like
+            The data transformed into shape as used in fit. The first
```
Same here, wrong docstring

kaichogami replied:
Other than `array-like` (should be array) the explanation seems correct no?

kingjr replied:
for inverse_transform.fit() we know that X is 2D with shape n_samples, n_features, you need to specify this in the docstring. Then the return is correct yes.

kaichogami commented on mne/decoding/transformer.py:
```diff
@@ -245,6 +246,111 @@ def inverse_transform(self, X, y=None):
+            The data to fit. Can be, for example a list, or an array of at
+            least 2d. The first dimension must be of length n_samples, where
+            samples are the independent samples used by the estimator
+            (e.g. n_epochs for epoched data).
```
@agramfort I have fixed the doc string of `inverse_transform`. I am not sure what is wrong with the doc string of `fit`.

kaichogami replied:
Wait.. When I view it from `conversation` it shows the commit where I have written `inverse_transform` doc string incorrectly. When I view it from `files changed` tab it is showing the latest commit.. Perhaps a bug with github?

kingjr replied:
you're probably looking at a particular commit. just refresh the PR? I'm seeing the wrong docstring either way

kingjr replied:
Ok, it's because we commented on a line below the chance. comment addressed.

kaichogami commented:
@agramfort @kingjr I have already addressed your comments [here](https://github.com/mne-tools/mne-python/pull/3409/commits/d56a30c5c85e3843a76ee339157fa66e14999407) since a long time. I am not sure why you are not able to see the changes.

kaichogami commented:
Github is likely showing the changes of some old commit rather than being updated.

kingjr commented:
you still have to say that X is 2D for inverse_transform. It's not at least 2D, it's just 2D, i.e. n_samples * n_features

kaichogami commented:
@kingjr I have done that. Please see the latest commit by going to thje commits tab
[picture](https://snag.gy/lONiyd.jpg)

kaichogami commented:
@kingjr @agramfort I apologize for getting you both confused about the doc string. 
Any more comments?
@dengemann

kingjr commented:
LGTM

agramfort commented on mne/decoding/tests/test_transformer.py:
```diff
@@ -160,3 +160,27 @@ def test_epochs_vectorizer():
+
+    orig_data = vect.inverse_transform(result)
+    assert_equal(orig_data.ndim, 3)
+    assert_array_equal(orig_data, data)
```
it does not seem you test inverse_transform with a different number of samples. this should pass. You could do:

assert_array_equal(vect.inverse_transform(result[1:]), data[1:])

ok?

agramfort commented on mne/decoding/transformer.py:
```diff
@@ -245,6 +247,108 @@ def inverse_transform(self, X, y=None):
+        -------
+        X : array
+            The data transformed into shape as used in fit. The first
+            dimension is of length (n_samples).
```
(n_samples) -> n_samples

agramfort commented on mne/decoding/transformer.py:
```diff
@@ -245,6 +247,108 @@ def inverse_transform(self, X, y=None):
+        Parameters
+        ----------
+        X : array-like, shape (n_samples,  n_features)
+            Transform data back to original shape.
```
Data to be transformed back to original shape.

agramfort commented on mne/decoding/transformer.py:
```diff
@@ -245,6 +247,108 @@ def inverse_transform(self, X, y=None):
+
+        Returns
+        -------
+        X : array, shape (n_samples, -1)
```
(n_samples, -1) -> (n_samples, n_features)

agramfort commented:
Thx
